### ISSUE TITLE
Performance Regression

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [microsoft/pyright#2960]
**Describe the bug**
This code block takes pyright 1.1.205 1 second. It takes pyright 1.1.206 3 seconds. It takes pyright 1.1.207+ about 30 seconds. The large speed drop makes pylance very slow. 

The two libraries that will need to be installed are tensorflow and numpy. I also use custom tensorflow type stubs as tensorflow lacks any types. Without using custom type stubs where tensorflow effectively becomes Any and performance is about 1 second. In the reproduction section I include a repository to help reproduce this.
 
```python
from __future__ import annotations

from typing import Callable, Iterable, List, Mapping, Optional, Sequence, Tuple, TypeVar, Union, cast
from typing_extensions import TypeGuard

import math
from collections import abc
from functools import partial

import numpy as np

import tensorflow as tf

T1 = TypeVar("T1")
T2 = TypeVar("T2")
ContainerGeneric = Union[Mapping[str, "ContainerGeneric[T1]"], Sequence["ContainerGeneric[T1]"], T1]
ContainerArrays = ContainerGeneric[Union[np.ndarray, np.number]]
ContainerTensors = ContainerGeneric[Union[tf.Tensor, tf.SparseTensor]]
ContainerAnyTensors = TypeVar("ContainerAnyTensors", ContainerTensors, ContainerArrays)
ArrayLike = Union[np.ndarray, np.number, tf.Tensor, tf.SparseTensor, float]

ComparisonResultT = Tuple[bool, Optional[List["int | str"]]]
DEFAULT_ABSOLUTE_TOLERANCE = 1e-4


def container_fmap(f: Callable[[T1], T2], elements: ContainerGeneric[T1]) -> ContainerGeneric[T2]:
    if isinstance(elements, abc.Mapping):
        return {key: container_fmap(f, val) for key, val in elements.items()}
    if isinstance(elements, tuple):
        return tuple(map(partial(container_fmap, f), elements))
    if isinstance(elements, abc.Sequence) and not isinstance(elements, (str, bytes)):
        return [container_fmap(f, elem) for elem in elements]
    return f(elements)


def _is_integer_dtype(dtype: np.dtype | tf.dtypes.DType) -> bool:
    if isinstance(dtype, np.dtype):
        return np.issubdtype(dtype, np.integer)
    else:
        return dtype.is_integer


def tensor_equality(
    tensor1: ArrayLike,
    tensor2: ArrayLike,
    atol: float = DEFAULT_ABSOLUTE_TOLERANCE,
    test_env: Optional[tf.test.TestCase] = None,
) -> bool:
    if isinstance(tensor1, tf.SparseTensor) or isinstance(tensor2, tf.SparseTensor):
        tensor1 = tf.sparse.to_dense(tensor1)
        tensor2 = tf.sparse.to_dense(tensor2)

    if isinstance(tensor1, float) and isinstance(tensor2, float):
        return math.isclose(tensor1, tensor2, abs_tol=atol)

    if isinstance(tensor1, (int, float)):
        if isinstance(tensor2, tf.Tensor):
            tensor1 = tf.constant(tensor1)
        else:
            tensor1 = np.array(tensor1)

    if isinstance(tensor2, (int, float)):
        if isinstance(tensor1, tf.Tensor):
            tensor2 = tf.constant(tensor2)
        else:
            tensor2 = np.array(tensor2)

    assert not isinstance(tensor1, (int, float))
    assert not isinstance(tensor2, (int, float))

    if type(tensor1) != type(tensor2):
        return False

    if tensor1.dtype != tensor2.dtype:
        return False

    if tensor1.shape != tensor2.shape:
        return False

    if test_env:
        assert isinstance(tensor1, tf.Tensor)
        array1 = test_env.evaluate(tensor1)
        array2 = test_env.evaluate(tensor2)
    else:
        if isinstance(tensor1, tf.Tensor) or isinstance(tensor2, tf.Tensor):
            array1 = tensor1.numpy()
            array2 = tensor2.numpy()
        else:
            array1 = tensor1
            array2 = tensor2

    if _is_integer_dtype(array1.dtype):
        return np.all(array1 == array2)
    else:
        comparisons = np.isclose(array1, array2, atol=atol, equal_nan=True)
        return comparisons.all()


def extract_nested_key(data: ContainerGeneric[T1], key_path: Iterable[int | str]) -> T1:
    result = data
    for key in key_path:
        if isinstance(key, int):
            assert isinstance(result, abc.Sequence)
            result = result[key]
        else:
            assert isinstance(result, abc.Mapping)
            result = result[key]

    return cast(T1, result)


def _tensor_collection_equality(
    tensors1: ContainerArrays,
    tensors2: ContainerArrays,
    atol: float = DEFAULT_ABSOLUTE_TOLERANCE,
    debug_path: Optional[List[str | int]] = None,
) -> ComparisonResultT:
    if debug_path is None:
        debug_path = []

    if not (isinstance(tensors1, type(tensors2)) or isinstance(tensors2, type(tensors1))):
        return (False, debug_path)

    if not isinstance(tensors1, (np.ndarray, dict, list, tuple, np.number)) or not isinstance(
        tensors2, (np.ndarray, dict, list, tuple, np.number)
    ):
        raise TypeError(f"Unexpected type for tensors1: {type(tensors1)}")

    if isinstance(tensors1, (np.ndarray, np.number)) or isinstance(tensors2, (np.ndarray, np.number)):
        result = tensor_equality(tensors1, tensors2, atol=atol)
        if result:
            return (result, None)
        else:
            return (result, debug_path)

    if len(tensors1) != len(tensors2):
        return (False, debug_path)

    if isinstance(tensors1, dict):
        assert isinstance(tensors2, dict)

        for key, value in tensors1.items():
            if key not in tensors2:
                return (False, debug_path + [key])
            key_result, key_debug_path = _tensor_collection_equality(
                value, tensors2[key], atol=atol, debug_path=debug_path + [key]
            )
            # Short circuit if possible.
            if not key_result:
                return (key_result, key_debug_path)
        return (True, None)

    assert isinstance(tensors2, (list, tuple))

    for key, (tensor1, tensor2) in enumerate(zip(tensors1, tensors2)):
        debug_extension: List[int | str] = [key]
        key_result, key_debug_path = _tensor_collection_equality(
            tensor1, tensor2, atol=atol, debug_path=debug_path + debug_extension
        )
        if not key_result:
            return (key_result, key_debug_path)

    return (True, None)


def evaluate_tensors(tensors: ContainerTensors, test_env: Optional[tf.test.TestCase]) -> ContainerArrays:
    if test_env:
        return test_env.evaluate(tensors)
    else:
        return container_fmap(lambda tensor: tensor.numpy(), tensors)  # type: ignore


def _log_tensor_equality_mismatch(
    arrays1: ContainerArrays, arrays2: ContainerArrays, debug_path: Iterable[str | int]
) -> None:
    debug_path_str = ",".join(map(str, debug_path))

    array1: np.ndarray | np.number
    array2: np.ndarray | np.number
    if debug_path_str != "":
        print(f"Tensor Collection mismatch occurred at {debug_path_str}")
        array1 = extract_nested_key(arrays1, debug_path)  # type: ignore
        array2 = extract_nested_key(arrays2, debug_path)  # type: ignore
    else:
        assert isinstance(arrays1, (np.ndarray, np.number))
        assert isinstance(arrays2, (np.ndarray, np.number))
        array1 = arrays1
        array2 = arrays2

    print(f"Mismatched tensors")
    print(array1)
    print(array2)

    shape_match = array1.shape == array2.shape
    dtype_match = array1.dtype == array2.dtype

    if not shape_match:
        print(f"Mismatch shapes")
        print(f"Shape 1: {array1.shape} Shape 2: {array2.shape}")

    if not dtype_match:
        print(f"Mismatch dtypes")
        print(f"Dtype 1: {array1.dtype} Dtype 2: {array2.dtype}")

    if shape_match and dtype_match:
        diff = np.absolute(array1 - array2)
        print(f"Maximum absolute difference: ", diff.max())
        print(f"Index Maximum Difference: ", np.unravel_index(diff.argmax(), diff.shape))


def check_container_tensors(tensors: ContainerTensors | ContainerArrays) -> TypeGuard[ContainerTensors]:
    if isinstance(tensors, (abc.Mapping, abc.Sequence)):
        if len(tensors) == 0:
            return True
        if isinstance(tensors, abc.Mapping):
            return check_container_tensors(next(iter(tensors.values())))
        else:
            return check_container_tensors(next(iter(tensors)))

    return isinstance(tensors, (tf.Tensor, tf.SparseTensor))


def tensor_collection_equality(
    tensors1: ContainerAnyTensors,
    tensors2: ContainerAnyTensors,
    atol: float = DEFAULT_ABSOLUTE_TOLERANCE,
    test_env: Optional[tf.test.TestCase] = None,
    log_debug_path: bool = True,
) -> bool:
    assert check_container_tensors(tensors1) == check_container_tensors(tensors2)

    if check_container_tensors(tensors1):
        arrays1 = evaluate_tensors(tensors1, test_env)
    else:
        arrays1 = cast(ContainerArrays, tensors1)

    if check_container_tensors(tensors2):
        arrays2 = evaluate_tensors(tensors2, test_env)
    else:
        arrays2 = cast(ContainerArrays, tensors2)

    result, debug_path = _tensor_collection_equality(arrays1, arrays2, atol=atol)

    if not result and log_debug_path:
        assert debug_path is not None
        _log_tensor_equality_mismatch(arrays1, arrays2, debug_path)

    return result
```

**To Reproduce**
I've made a small [repository](https://github.com/hmc-cs-mdrissi/pyright_generic_perf) with the code + necessary type stubs/environment to reproduce. If you do pip install -r requirements.txt in the repo you should have needed environment. Tensorflow has no type stubs so I've started writing my own and these stubs are necessary to trigger slowness. Before when tensorflow types were treated like Any I didn't see an issue.

**Expected behavior**
No performance drop.

**Comments**
pyright --verbose --stats is very useful. This allowed me to identify where the bottleneck was coming from pretty quickly.

**Comment by erictraut** [comment]
Thanks for the bug report and the repro. The performance regression is due to this seemingly-innocuous one-line change: https://github.com/microsoft/pyright/commit/844f7cb98987955dc617cd97b1372325e76a4530. It's due to the way you're using a recursive type alias. I'll need to investigate further to see if there's a way I can make the performance better in this case without breaking other cases (including the one that prompted the change that triggered this regression).

**Comment by hmc-cs-mdrissi** [comment]
If there's an alternative way to write the types that is performant, I'm fine adjusting them. These types are probably about the most complex ones I have in my codebase.

I'd also expect two commits to be relevant. 1.1.206 -> 1.1.207 had most of performance change, but 1.1.205 -> 1.1.206 also had speed drop that's consistently measurable. Although 1 vs 3 seconds is more acceptable change.

**Comment by erictraut** [comment]
The changes that occurred between 1.1.205 and 1.1.207 were not the root cause of the slowdown. They exacerbated the underlying cause. I've found and fixed the root cause, so performance is now much better when analyzing recursive types. Analysis time for your sample file went from about 32sec on my machine to about 325ms — about a 100x speedup. This improvement will be in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.217, which I just published. It will also be included in the next release of pylance.
